### PR TITLE
Set and delete using name of HW

### PR DESCRIPTION
**Pull request by AnishReddyRavula**
Rename command line flag `--mgmt_addr` to `--management_address`
`set` and `delete` uses `list` to work with name_or_uuid in the command line

### CODE REVIEW

File: doniclient/v1/client.py [hunk 1]
```diff
@@ -50,11 +50,26 @@ def create(self, json, **kwargs):
...
+        hardware_list = self.list()
+        matching_hardware = [h for h in hardware_list if h['name'] == name_or_uuid]
+        hardware_uuid = name_or_uuid
+        if matching_hardware:
+            hardware_uuid = matching_hardware[0]['uuid']
+        resp = self.adapter.patch(f"/v1/hardware/{hardware_uuid}/", json=json)
+        try:
+            return resp.json()
+        except json.JSONDecodeError:
+            return resp
```
Comment (msherman64) [inline doniclient/v1/client.py:75]: I would prefer if we kept the methods in `/v1/client.py` a relatively 1-1 match to the REST API, as other python code may import this library to use that.

Instead, we can add the extra logic to the CLI methods under `osc`.

For example, replacing:
`hw_client.delete(parsed_args.uuid)`
with
```
# find_exactly_one should throw an error if more than one match is found.
hardware_instance = find_exactly_one(hw_client, name_or_uuid)
hw_client.delete(hardware_instance.uuid)
```

**Review by msherman64 — CHANGES_REQUESTED** [review]
Logic is overall OK, but requesting a few refactors:

1. Handle the "name_or_uuid" logic in the osc/cli module, rather than in api/v1/client.
2. When listing hardware, you should be able to set a "filter" argument, rather than doing a list comprehension. This will speed up the query.
3. throw an error if more than one result is returned, behavior is undefined in this case. (We can't safely just choose item [0])
4. lets move the "mgmt_iface" fix to it's own commit and PR for clarity, but you can "chain"  this pr off of that one.

File: doniclient/osc/cli.py [hunk 2]
```diff
@@ -56,7 +56,17 @@ def take_action(self, parsed_args):
...
+        worker_types = set([j['worker_type'] for i in data for j in i['workers']])
+        print(worker_types)
+        print(labels)
+        pep = (
+            oscutils.get_dict_properties(
+                    s, columns, formatters={"Properties": oscutils.format_dict}
+            )
+            for s in data
+        )
+        print(i for i in pep)
```
Comment (msherman64) [inline doniclient/osc/cli.py:69]: These look like debugging commands for the next PR?

File: doniclient/osc/cli.py [hunk 3]
```diff
@@ -163,7 +179,7 @@ def get_parser(self, prog_name):
             parser,
             "baremetal",
             {
-                "mgmt_addr": PropertyFlag("mgmt_addr", str, None),
+                "management_address": PropertyFlag("management_address", str, None),
```
Comment (msherman64) [inline doniclient/osc/cli.py:182]: This should disappear from this PR after rebasing onto the other commit

File: doniclient/tests/osc/test_hardware.py [hunk 4]
```diff
@@ -1,31 +1,37 @@
...
     (
         "baremetal",
-        "--mgmt_addr",
-        "mgmt_addr",
-        "/properties/mgmt_addr",
+        "--management_address",
+        "management_address",
+        "/properties/management_address",
         "fake-mgmt_addr",
+        True,
```
Comment (msherman64) [inline doniclient/tests/osc/test_hardware.py:16]: Same as other comment, I'm assuming that a rebase will remove from this patch-set.

File: doniclient/tests/osc/test_hardware.py [hunk 5]
```diff
@@ -126,19 +132,75 @@ class TestHardwareDelete(TestHardware):
...
+        self.hardware_mock.find.side_effect = Exception
+        # return two hardwares with same name
+        self.hardware_mock.list.return_value = [
+            hardware_fakes.FakeHardware.create_one_hardware(),
+            hardware_fakes.FakeHardware.create_one_hardware()
+        ]
+        try:
+            self.cmd.take_action(parsed_args)
+        except exceptions.CommandError as e:
+            self.assertEqual(e.args[0], f"More than one resource exists with the name or ID '{FAKE_HARDWARE_NAME}'.")
```
Comment (msherman64) [inline doniclient/tests/osc/test_hardware.py:152]: Nice! I'm hoping we can reuse this general pattern to add better tests to Blazar as well.
👍

File: doniclient/tests/osc/test_hardware.py [hunk 6]
```diff
@@ -148,15 +210,17 @@ def test(self):
                 assert parsed_args.properties == {prop: value}
 
                 self.cmd.take_action(parsed_args)
+                # test if the get is called with name
+                self.hardware_mock.get.assert_called_with(name_or_id)
                 self.hardware_mock.update.assert_called_with(
                     FAKE_HARDWARE_UUID, [{"op": "add", "path": path, "value": value}]
```
Comment (msherman64) [inline doniclient/tests/osc/test_hardware.py:216]: Good to see that this still passes with our wrapper to look up by name. LGTM!

**Review by msherman64 — APPROVED** [review]
LGTM, approved after a rebase onto the other PR to clean this one up, and removing the cli columns bit.

File: doniclient/osc/cli.py (hunk 2 above)
Comment (AnishReddyRavula) [reply doniclient/osc/cli.py:69]: Ah. right.

**Pull request merged**